nose2/plugins/prettyassert.py

```diff
@@ -0,0 +1,360 @@
+            return
+
+        """
+        CASE 1: Operator token
+
+        skip tokens and either leave resolution in progress or reset, depending
+
+        continue resolution for
+          "."
+            because that's what attribute resolution *is*
+          ")"
+            this is handy, as it means that "(x).y" works
+
+        reset resolution for everything else, e.g. "[", "]", ":"
+        special note: reset resolution for "("
+
+        failing to filter out ")" can result in badness in cases like this:
+            >>> def foo():
+            >>>     return [1]
+            >>> foo.pop = 2
+            >>> ...
+            >>> def test_foo():
+            >>>    assert foo().pop() == 2
+
+        if we stop resolution when we see an LPAREN, we resolve `foo`
+        successfully, fail on `pop` and everything is OK, but if we try to
+        traverse the LPAREN, we get `foo.pop = 2` in our values, which is
+        wrong
+        """
+        if toktype == tokenize.OP:
+            if tok not in (".", ")"):
+                self.doing_resolution = None
+            return
+
+        # CASE 2: "assert" statement
+        # assert statement was reached, reset
+        # return the start line (start = (startrow, startcol))
+        if tok == 'assert':
+            self.seen_tokens.clear()
+            self.doing_resolution = None
+            return start[0]
+
+        # handle tokens
+
+        # CASE 3: a name is being resolved,
+        #         there is a previous token,
+        #         and it's a "." operator
+        if self.doing_resolution and prior_tok and (
+                prior_tok[0] == tokenize.OP and prior_tok[1] == '.'):
+            # unpack and look for the attribute
+            obj, name = self.doing_resolution
+            if hasattr(obj, tok):
+                obj = getattr(obj, tok)
+                name = name + '.' + tok
+                self.doing_resolution = (obj, name)
+                self.seen_tokens[name] = obj
+            # if we couldn't find a relevant attribute, reset on resolution so
+            # that we can try afresh
+            else:
+                self.doing_resolution = None
+
+        # CASE 4: a name is being resolved and there is no preceding "." or
+        #         resolution was explicitly stopped
+        else:
+            # skip tokens we've seen, but grab them as the current things under
+            # resolution
+            if tok in self.seen_tokens:
+                self.doing_resolution = (self.seen_tokens[tok], tok)
+                return
+            # we've never seen this token before
+            else:
+                # try to resolve to a value
+                try:
+                    value = self.f_locals[tok]
+                except KeyError:
+                    try:
+                        value = self.f_globals[tok]
+                    except KeyError:
+                        # unresolveable name -- short circuit
```

Comment:
What situations could lead to this and what is the overall outcome of returning here?

Reply:
The outcome is that we skip this value in the output.
Simple example scenario:
```
x = 1
assert x == 0; y = 1
```

We'll hit `y` and not have a value for it in the frame. It should be omitted from output.

I'll add a test case for that when I get to all of this (proving harder than expected).